### PR TITLE
[RFR] Better integration with module bundlers

### PR DESCRIPTION
**Pull request by jpetitcolas**
- [x] Update README and UPGRADE files
- [x] Upgrade Getting started snippets
- [x] Update `package.json` to point on correct files

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,9 +22,42 @@ npm install ng-admin --save
...
+
+``` html
+<link rel="stylesheet" href="node_modules/ng-admin/build/ng-admin.min.css">
+<script src="node_modules/ng-admin/build/ng-admin.min.js"></script>
+```
+
+## Bootstraping your Admin
+
+Just add a `<div ui-view>` into your HTML, add a `ng-app` attribute on your `body`, and configure the admin:
+ithu
```
Comment (jpetitcolas) [inline README.md]: Seems I missed an `ALT + TAB`. To delete.

File: UPGRADE-1.0.md [hunk 2]
```diff
@@ -39,3 +39,25 @@ After editing an entity, ng-admin now redirects to the previous page (e.g. list
...
+```
+
+Furthermore, if your module bundler supports CSS and/or SASS, you can also embed styles using:
+
+``` js
+// SASS version
+require('ng-admin/sass/ng-admin.scss');
+
+// CSS version
+require('ng-admin/build/ng-admin.css');
```
Comment (jpetitcolas) [inline UPGRADE-1.0.md]: `ng-admin.min.css`

File: doc/Getting-started.md [hunk 3]
```diff
@@ -36,7 +38,35 @@ cd my-first-admin
...
+
+### With a Module Bundler
+
+If you use a module bundler such as Webpack, you can use `ng-admin` using a simple `require`:
+
+``` js
+require('ng-admin');
+```
+
+Including `ng-admin` this way would also ng-admin SASS files to generate styles. Hence, you would need
```
Comment (jpetitcolas) [inline doc/Getting-started.md]: Remove this paragraph and add the requirement of either the SASS or CSS version in previous snippet.

File: package.json [hunk 4]
```diff
@@ -66,7 +66,7 @@
     "sinon": "~1.14.1",
     "style-loader": "^0.12.2",
     "superagent": "^0.18.2",
-    "textangular": "1.4.3",
+    "textangular": "~1.5.1",
```
Comment (jpetitcolas) [inline package.json]: This dependency (and a lot of others) are no more dev. Move them accordingly.

File: src/javascripts/vendors.js [hunk 5]
```diff
@@ -2,11 +2,13 @@ global.rangy = require('../../node_modules/rangy/lib/rangy-core');
 global.rangy = require('../../node_modules/rangy/lib/rangy-selectionsaverestore');
 global.numeral = require('numeral');
 
+require('angular');
 require('angular-ui-router');
 require('../../node_modules/textangular/dist/textAngular-sanitize');
 require('angular-ui-codemirror');
+require('textangular/dist/textAngular-sanitize'),
+require('imports?define=>false!textangular');
```
Comment (jpetitcolas) [inline src/javascripts/vendors.js]: Ugly, but required in order to imports textAngular correctly.

File: src/sass/ng-admin.scss [hunk 6]
```diff
@@ -1,3 +1,18 @@
...
+@import '../../node_modules/bootstrap-sass/assets/stylesheets/bootstrap.scss';
+@import '~nprogress/nprogress.css';
+@import '~humane-js/themes/flatty.css';
+@import '~textangular/src/textAngular.css';
+@import '~codemirror/lib/codemirror.css';
+@import '~codemirror/addon/lint/lint.css';
+@import '~ui-select/dist/select.css';
+@import './sb-admin.scss';
+
+// require('ng-admin/build/ng-admin.min.css');
```
Comment (jpetitcolas) [inline src/sass/ng-admin.scss]: Remove

**Comment by jpetitcolas** [comment]
With the help of @RobinBressan, switching to RFR.

@djhi, meanwhile, can you try using this branch on your project? I heard you got some troubles building an admin app in production with master version of ng-admin. It should fix your issue.

**Comment by Phocea** [comment]
Hello,

I am actually in the process of using the ng-admin-only way because I needed to add ngCookies  to my app.
As you stated, the production page is out of date and it was a painful process to get it right again. I have now something working but I am not 100% that ll the dependencies I added are needed!

Anyway, does this branch means that it will be solved by simply adding a require(ng-admin) in the js loaded by our webpack build, as well as to ngCookies. And will simply be able to declare the module when initializing the module ?

`var myApp = angular.module('myApp', [ 'ng-admin','ngCookies' ]);`

If so, I will put my refactor on the backboiler for now, and wont submit my changes for the production documentation :)

Thanks

**Comment by jpetitcolas** [comment]
@Phocea: unfortunately for your changes, this is indeed the aim of this PR. :)

File: package.json (hunk 4 above)
Comment (djhi) [inline package.json]: We should fix all package versions currently prefixed with the evil `^`

Comment (djhi) [inline package.json]: And maybe avoid the `~` too

File: README.md [hunk 7]
```diff
@@ -22,9 +22,42 @@ npm install ng-admin --save
...
+
+``` js
+require('ng-admin');
+```
+
+If your module bundler also supports SASS or CSS, you can also include stylesheets using:
+
+``` js
+// SASS version
+require('ng-admin/sass/ng-admin.scss');
```
Comment (djhi) [inline README.md]: s/ng-admin/sass/ng-admin.scss/ng-admin/src/sass/ng-admin.scss

**Comment by jpetitcolas** [comment]
To sum up a private discussion with @djhi:

- [ ] Move `admin-config` as a prod dependency
- [ ] Update Webpack configuration sample to exclude node_modules

**Comment by Phocea** [comment]
@jpetitcolas great :) Well this is great news really because the old way was really painful and as you stated, the production.md page was impossible to maintain.
Can't wait to have this available on main. Thanks for the rework

**Comment by Phocea** [comment]
Hello @jpetitcolas , any update on when this is going to be merged on the main branch? No rush, but it would sort out the mess I had to implement :)
Thanks

**Comment by jpetitcolas** [comment]
@Phocea you're right, this PR definitively needs some love. I take back a look on it. :)

**Comment by Phocea** [comment]
Merci Monsieur !

File: README.md [hunk 8]
```diff
@@ -22,9 +22,42 @@ npm install ng-admin --save
...
+``` js
+// SASS version
+require('ng-admin/src/sass/ng-admin.scss');
+
+// CSS version (do not include minified version for better performances)
+require('ng-admin/build/ng-admin.css');
+```
+
+Using a module bundler, you would also be able to generate the source map for all your JavaScript
+and stylesheets, helping you to hunt even the most pernicious bugs.
```
Comment (fzaninotto) [inline README.md]: s/pernicious/obscure/

File: README.md [hunk 9]
```diff
@@ -22,9 +22,42 @@ npm install ng-admin --save
...
+
+### Without a Module Bundler
+
+If you don't have a module bundler, don't worry: you can still embed `ng-admin` the old way:
```
Comment (fzaninotto) [inline README.md]: s/embed `ng-admin` the old way/include `ng-admin` with a `<script>` tag/

File: UPGRADE-1.0.md [hunk 10]
```diff
@@ -39,3 +39,25 @@ After editing an entity, ng-admin now redirects to the previous page (e.g. list
...
+// SASS version
+require('ng-admin/sass/ng-admin.scss');
+
+// CSS version
+require('ng-admin/build/ng-admin.min.css');
+```
+
+## Removing `ng-admin-only` build
+
+We used to build two versions of `ng-admin`: a standalone one and a `only` one. The latter includes only files required to ng-admin. But it was a little buggy and requires a lot of dependencies. Including a functional version of this only version was so painful we decided to abandon it in profit of a better bundler integration (see previous paragraph).
```
Comment (fzaninotto) [inline UPGRADE-1.0.md]: s/and requires/and required/

Comment (fzaninotto) [inline UPGRADE-1.0.md]: s/this only version/this '-only' version/

Comment (fzaninotto) [inline UPGRADE-1.0.md]: s/painful we/painful that we/

File: doc/Getting-started.md [hunk 11]
```diff
@@ -9,6 +9,8 @@ Let's use the REST API offered by [JSONPlaceholder](http://jsonplaceholder.typic
```
Comment (fzaninotto) [inline doc/Getting-started.md]: I think you should revert the changes on this one, to let newcomers begin with the simplest version (`<script>` tag)

File: package.json [hunk 12]
```diff
@@ -19,9 +21,14 @@
...
-    "babel-core": "^5.2.17",
-    "babel-loader": "^5.0.0",
+    "babel": "^6.5.2",
+    "babel-cli": "^6.14.0",
+    "babel-core": "^6.14.0",
+    "babel-loader": "^6.2.5",
+    "babel-plugin-add-module-exports": "^0.2.1",
+    "babel-preset-es2015": "^6.14.0",
+    "babel-preset-stage-1": "^6.13.0",
+    "babel-preset-stage-2": "^6.13.0",
```
Comment (jpetitcolas) [inline package.json]: To remove

File: package.json [hunk 13]
```diff
@@ -19,9 +21,14 @@
...
-    "babel": "^4.6.0",
-    "babel-core": "^5.2.17",
-    "babel-loader": "^5.0.0",
+    "babel": "^6.5.2",
+    "babel-cli": "^6.14.0",
+    "babel-core": "^6.14.0",
+    "babel-loader": "^6.2.5",
+    "babel-plugin-add-module-exports": "^0.2.1",
+    "babel-preset-es2015": "^6.14.0",
+    "babel-preset-stage-1": "^6.13.0",
```
Comment (jpetitcolas) [inline package.json]: To remove

File: src/sass/ng-admin.scss [hunk 14]
```diff
@@ -255,12 +268,9 @@ div.bottom-loader {
     }
 
     .remove_button_container {
-        position: absolute;
-        right: -7em;
-        z-index: 2;
+        float: right;
```
Comment (jpetitcolas) [inline src/sass/ng-admin.scss]: Use master version

File: README.md [hunk 15]
```diff
@@ -22,9 +22,42 @@ npm install ng-admin --save
...
+// CSS version (do not include minified version for better performances)
+require('ng-admin/build/ng-admin.css');
+```
+
+Using a module bundler, you would also be able to generate the source map for all your JavaScript
+and stylesheets, helping you to hunt even the most obscure bugs.
+
+### Without a Module Bundler
+
+If you don't have a module bundler, don't worry: you can still embed `ng-admin` the old way:
```
Comment (fzaninotto) [inline README.md]: s/embed `ng-admin` the old way/include `ng-admin` with a `<script>` tag/

File: UPGRADE-1.0.md [hunk 16]
```diff
@@ -39,3 +39,28 @@ After editing an entity, ng-admin now redirects to the previous page (e.g. list
...
+// SASS version
+require('ng-admin/lib/sass/ng-admin.scss');
+
+// CSS version
+require('ng-admin/build/ng-admin.min.css');
+```
+
+## Removing `ng-admin-only` build
+
+We used to build two versions of `ng-admin`: a standalone one and a `only` one. The latter includes only files required to ng-admin. But it was a little buggy and required a lot of dependencies. Including a functional version of this only version was so painful we decided to abandon it in profit of a better bundler integration (see previous paragraph).
```
Comment (fzaninotto) [inline UPGRADE-1.0.md]: s/this only version/this '-only' version/ s/painful we/painful that we/

**Comment by jpetitcolas** [comment]
Code review applied.

**Comment by fzaninotto** [comment]
\o/

**Comment by jpetitcolas** [comment]
image

**Comment by Phocea** [comment]
+1 .... here we go, more work for me to come ;)
As this method been applied to the demo by any chance ?

**Comment by jpetitcolas** [comment]
@Phocea: good point: we forget that. Need to work on it.

However, if you speak of Poster Galore, here is the related PR: https://github.com/marmelab/ng-admin-demo/pull/26

**Comment by Phocea** [comment]
@jpetitcolas  Yeap meant the poster galore demo . Sorry more work for you too then :)

**Comment by Phocea** [comment]
Hello @jpetitcolas,

I have started to use your refactor in our own webpack build and went through a bit of trouble :)
I pinpointed it to a path problem, but kept going in circle about it and could not see a solution without making changes to ng-admin structure itself.
Let me try to explain:

**1.**
```
var myApp = angular.module('myApp', [
 require('ng-admin'),
]);
```
The only way for webpack to find the ng-admin module is to either give a path relative to the javascript this is required from, or add an alias in webpack.config like so (for example):

```
  resolve : {
    root : [ path.join(__dirname, 'node_modules') ],
    alias : { 'ng-admin' : 'ng-admin/src/javascripts/ng-admin.js',}
  },
```
Might be something to add to the documentation ?

**2.**
With above fixed, I am now getting some errors when webpack is resolving the require statements present in ng-admin vendor.js. All which refer to a package directly inside the node_modules is ok, all the other are failing to find the related file.
Would it not be better to use aliases in the ng-admin webpack config to avoid that path configuration trouble.
Also the vendor.js looks like it could be simplified:

- No need for path for angular.
- textAngular-sanitize mentionned twice

In the end I have managed to have a it wokring with the following resolve section in my webpack config:
```
  resolve : {
    root : [ path.join(__dirname, 'node_modules') ],
    alias : { 'ng-admin' : 'ng-admin/src/javascripts/ng-admin.js',
              'rangy-core' : 'rangy/lib/rangy-core',
              'rangy-selectionsaverestore' : 'rangy/lib/rangy-selectionsaverestore',
              'textAngular-sanitize' : 'textangular/dist/textAngular-sanitize',
              'angular-numeraljs' : 'angular-numeraljs/dist/angular-numeraljs',
              'ui-bootstrap-tpls' : 'angular-ui-bootstrap/dist/ui-bootstrap-tpls',
              'ng-file-upload' : 'ng-file-upload/dist/ng-file-upload',
            }
  },
```
and ng-admin vendor.js modified to look like:

```
global.rangy = require('rangy-core');
global.rangy = require('rangy-selectionsaverestore');
global.numeral = require('numeral');

require('angular');
require('angular-ui-router');
require('angular-ui-codemirror');
require('textAngular-sanitize'),
require('textangular');
require('nginflection');
require('ui-select');
require('angular-translate');

require('angular-numeraljs');
require('ui-bootstrap-tpls');
require('ng-file-upload');

global._ = require('underscore');
```

Does this sound acceptable ?

**Comment by jpetitcolas** [comment]
@Phocea: there was some mistakes in this build (we didn't see them because of the use of `npm link`). Can you try with the updated branch on https://github.com/marmelab/ng-admin/pull/1189? 

I'm on [Gitter](https://gitter.im/marmelab/ng-admin) if you need some support. But I'm working on a brand new project using the branch related to my PR, and it seems to work.

**Comment by Phocea** [comment]
lol .. was just about to say that I have a bit of css problem now !
Will try to grab your branch.. ! am building on windows using Maven (!!) so may encounteer some unexpected stuff.
I really have to stop making the same fixes you do ;)
